### DISCUSSION TITLE
VDDIN_OC_LIMIT cannot be set

### DISCUSSION THREAD

**Discussion by chhg82** [OE4T/meta-tegra#1506] [Q&A] [answered]
Hi,
I'm using kirkstone/JetPack5.1.2 on a Xavier NX device. When setting the power mode `sudo nvpmodel -m 8` an error occurs that apparently `VDDIN_OC_LIMIT` cannot be set.
If I comment out the following in `/etc/nvpmodel.conf` under power model definitions it works:

> #VDDIN_OC_LIMIT WARN 4000
#VDDIN_OC_LIMIT CRIT 5000

The limits are set here:

> < PARAM TYPE=HWMON NAME=VDDIN_OC_LIMIT >
WARN /sys/devices/platform/c250000.i2c/i2c-7/7-0040/hwmon curr1_max
CRIT /sys/devices/platform/c250000.i2c/i2c-7/7-0040/hwmon curr1_crit

It seems that `hwmon` is missing on my NX:

> XavierNX:~$ ll /sys/devices/platform/c250000.i2c/i2c-7/7-0040/
total 0
-r--r--r-- 1 root root 4096 Mar 19 15:36 modalias
-r--r--r-- 1 root root 4096 Mar 19 15:19 name
lrwxrwxrwx 1 root root    0 Mar 19 15:36 of_node -> ../../../../../firmware/devicetree/base/i2c@c250000/ina3221@40
drwxr-xr-x 2 root root    0 Mar 19 15:29 power
lrwxrwxrwx 1 root root    0 Apr 28  2022 subsystem -> ../../../../../bus/i2c
-rw-r--r-- 1 root root 4096 Apr 28  2022 uevent

Which kernel module loads hwmon and what could be the problem that it is not present for me?
Thank you!

**Comment by chhg82** [comment]
I found out that if I generate linux-tegra with `CONFIG_SENSORS_INA3221=y` and thus compile the power state chipset driver into the kernel, then `hwmon` is linked and `nvpmodel -m 8` runs! Why is `CONFIG_SENSORS_INA3221=m` (by default) not working as kernel module and what is the best practice?

**Reply by madisongh** [reply]
> Why is CONFIG_SENSORS_INA3221=m (by default) not working as kernel module 

That might depend on how you've built your images.  We include `kernel-module-ina3221`  in `MACHINE_EXTRA_RRECOMMENDS` in the machine config, and we have `ina3221` in the `KERNEL_MODULE_AUTOLOAD` setting, so it should get loaded at boot time.  If you've modified/overridden those settings, for instance, that could cause this problem.  Other customizations might do this, too.

**Reply by chhg82** [reply]
My custom machine is based on `jetson-xavier-nx-devkit-emmc.conf`, so as you say, `MACHINE_EXTRA_RRECOMMENDS` and `KERNEL_MODULE_AUTOLOAD` are already set correctly. However, in `tegra-nvpmodel-base.bbappend` an `RDEPENDS:${PN} += "kernel-module-ina3221"` seems to be necessary for the module to be loaded. What do you think about that?

**Reply by madisongh** [reply]
Kernel module/driver dependencies use `RRECOMMENDS`  rather than `RDEPENDS`; otherwise, if the driver is built into the kernel (`y` instead of `m`), the dependency will fail.

If you're building an image that excludes `packagegroup-machine-base` (for example, `core-image-minimal` or something based off that recipe), you won't get the `MACHINE_EXTRA_RRECOMMENDS` packages - that's the packagegroup that includes them.

**Reply by chhg82** [reply]
Yes thanks, `RRECOMMENDS` does it too!

You're right, I'm using `core-image-minimal`. I was wondering because `bitbake -e myimage` has the following output:

> MACHINE_EXTRA_RRECOMMENDS="kernel-module-nvgpu kernel-module-snd-hda-tegra kernel-module-snd-hda-codec-hdmi kernel-module-ina3221 kernel-module-tegra-bpmp-thermal kernel-module-spi-tegra114 kernel-module-ahci kernel-module-max77620-thermal kernel-module-pwm-fan kernel-module-tegra-udrm kernel-module-rtk-btusb kernel-module-rtl8822ce tegra-firmware-rtl8822"

...but at runtime only the `nvgpu` and `nvmap` modules are loaded (now also `ina3221`).

Wouldn't it make sense to integrate it into `tegra-nvpmodel-base.bb` in `meta-tegra`? Could there be any other potentially necessary `RRECOMMENDS` that I haven't noticed yet?

**Reply by madisongh** [reply, marked as answer]
Yes, pretty much everything listed in the MACHINE_EXTRA_RRECOMMENDS and MACHINE_EXTRA_RDPENDS variables.  Really, I don't recommend using `core-image-minimal` as the starting point for anything where you actually want to use the features of the hardware - as the Yocto documentation states, that recipe just gets you the bare minimum to boot the system.  `core-image-base`  is a better base to work from.
